js/renderer.ts

```diff
@@ -852,11 +872,11 @@ export abstract class ReactUiBase extends Component implements ReactComp {
         const s = bottomRight[0] - topLeft[0];
         this.object.setScalingLocal([s, s, s]);
         /* Convert from yoga units to 0-1 */
-        this.width = this.engine.canvas.clientWidth;
-        this.height = this.engine.canvas.clientHeight;
+        this.dpr = window.devicePixelRatio ?? 1;
```

Comment:
`window.devicePixelRatio` is always defined (on the browser), so no need to default it, I think? You should be able to use it directly everywhere, though I'm unsure what would happen on DPI changes (e.g. dragging a browser window from a DPI screen to non-DPI screen)

Reply:
> window.devicePixelRatio is always defined (on the browser), so no need to default it, I think? 
I checked the specs, it is since 2013. So I think we're safe to assume it's always there. I'll remove the `??`

> I'm unsure what would happen on DPI changes
I've actually tested this with changing the device in the devtools. As the resolution changes, the resize event gets triggered and the DPR is updated.